### PR TITLE
docker: clickhouse-server on the top of busybox

### PR DESCRIPTION
**Pull request by filimonov**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Build/Testing/Packaging Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Attempt to create a docker image to run clickhouse on the top of busybox instead of ubuntu. 

Detailed description / Documentation draft:
experimental / was not tested.

Uses glibc components from ubuntu 20.04.

Actually it questionable if we should now migrate to busybox (since lot of docker users already rely on ubuntu features of container, and since the final container size in only slightly smaller).

### CODE REVIEW

File: docker/server-busybox/root/entrypoint.sh [hunk 1]
```diff
@@ -0,0 +1,152 @@
+#!/bin/sh
```
Comment (filimonov) [inline docker/server-busybox/root/entrypoint.sh:1]: several changes done to work on sh

**Comment by alexey-milovidov** [comment]
LGTM.

Motivation (vague):
- I've met some people that tend to say that using Ubuntu for Docker images is bad practice;
- when we base our images on Ubuntu, container scanners report about old software unrelated to ClickHouse.

We will merge as soon as there will be any evidence that it's working.
(It will not replace the official container but we can start to experiment)

**Comment by filimonov** [comment]
Well i've made 2 images this way:
filimonovq/clickhouse-server:20.8.4.11-busybox
filimonovq/clickhouse-server:20.9.3.45-busybox

They work and pass some basic tests.  The main problems - lack of bash (so if you pass some scripts in you need to have proper shebang),  and lack of any package manager (in contrast to alpine, so you can't install anything inside with simple commands). 

The size difference: 
20.9.3.45  143.01 MB (busybox) 182.81 MB (current)
20.8.4.11  142.56 MB (busybox) 182.37 MB (current)

Maybe we should also try alpine + glibc added (it gives more flexibility, and maybe one day we will be able to switch to musl)?

**Comment by filimonov** [comment]
It looks like alpine version also ok,  i've pushed the image here: 
* filimonovq/clickhouse-server:20.9.3.45-alpine 

it's about 2Mb bigger than busybox as expected, but has lot advantages (better with security, package manager, will be the natural choice if we will ever switch to musl). 

patch to make it on the top of alpine:

```diff
diff --git a/docker/server-busybox/Dockerfile b/docker/server-busybox/Dockerfile
index e6426b6e56..dbe10766e5 100644
--- a/docker/server-busybox/Dockerfile
+++ b/docker/server-busybox/Dockerfile
@@ -1,4 +1,4 @@
-FROM busybox
+FROM alpine
 
 ENV LANG=en_US.UTF-8 \
     LANGUAGE=en_US:en \
@@ -15,7 +15,8 @@ RUN addgroup clickhouse \
     && chmod 700 /var/lib/clickhouse \
     && chown root:clickhouse /var/log/clickhouse-server \
     && chmod 775 /var/log/clickhouse-server \
-    && chmod +x /entrypoint.sh
+    && chmod +x /entrypoint.sh \
+    && apk add --no-cache su-exec
 
 EXPOSE 9000 8123 9009
 
diff --git a/docker/server-busybox/root/entrypoint.sh b/docker/server-busybox/root/entrypoint.sh
index 9c59fbb73b..e2edda9ca2 100755
--- a/docker/server-busybox/root/entrypoint.sh
+++ b/docker/server-busybox/root/entrypoint.sh
@@ -14,7 +14,7 @@ if [ "$(id -u)" = "0" ]; then
     USER=$CLICKHOUSE_UID
     GROUP=$CLICKHOUSE_GID
     # busybox has setuidgid & chpst buildin
-    gosu="setuidgid $USER:$GROUP"
+    gosu="su-exec $USER:$GROUP"
 else
     USER="$(id -u)"
     GROUP="$(id -g)"
```

How did i do those builds:
```
$ pwd
/home/mfilimonov/workspace/ClickHouse-docker-minimal/docker/server-busybox
git clean -f -X  ./
rm -rf ./ClickHouse
./download_tgz.sh && ./prepare.sh && docker build . -t filimonovq/clickhouse-server:20.9.3.45-alpine --pull 
docker login
docker push filimonovq/clickhouse-server:20.9.3.45-alpine 
```

**Comment by alexey-milovidov** [comment]
Ok.

PS. We want to switch to Musl, the main blocker is the need to embed .so loader into our binary to load:
- library dictionary source;
- catboost.

**Comment by filimonov** [comment]
Question - does clickhouse relies somehow on locales? How can it be tested? Any other potential issues which we should recheck?

**Comment by alexey-milovidov** [comment]
> Question - does clickhouse relies somehow on locales?

No. 

clickhouse-client may work bad if `C` locale is used instead of `UTF-8`, but it's not the issue of clickhouse-client itself but the environment.

There is a risk that some external libraries depend on locales, but in 99% it is erroneous, developers may think that locales are for good and use them without any reason (and they will do harm in non-obvious cases).

**Comment by filimonov** [comment]
But we need to ensure the clickhouse-client can work properly from server docker too.
What are potential issues with client? 

It seems it was added by intent:
https://github.com/ClickHouse/ClickHouse/commit/6e8f4c6fecf17d8b2fa6175e043e856461c60181

**Comment by alexey-milovidov** [comment]
`LC_ALL=C screen`
`clickhouse-client`

hangs while displaying progress bar.

**Review by alexey-milovidov — APPROVED** [review]

**Comment by filimonov** [comment]
> `LC_ALL=C screen`
> `clickhouse-client`
> 
> hangs while displaying progress bar.

Checked that. Indeed i see some issues with clickhouse-client in that scenario, but defaults in that alpine container work good.

**Comment by filimonov** [comment]
BTW: many alpine-based images still have bash (for better scripting in init files / entrypoint).

**Comment by alexey-milovidov** [comment]
@filimonov Should we merge busybox or Alpine based image? It's up to your choice.

**Comment by filimonov** [comment]
Alpine.

**Comment by filimonov** [comment]
Superseded by https://github.com/ClickHouse/ClickHouse/pull/16190

**Pull request closed without merging**